SimTracker/SiStripDigitizer/plugins/SiStripApvSimulationParametersESProducer.cc

```diff
@@ -0,0 +1,121 @@
+#include "FWCore/Framework/interface/ESProducer.h"
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "CondFormats/DataRecord/interface/SiStripCondDataRecords.h"
+#include "CondFormats/SiStripObjects/interface/SiStripApvSimulationParameters.h"
+#include <fstream>
+#include <boost/range/adaptor/indexed.hpp>
+
+class SiStripApvSimulationParametersESProducer : public edm::ESProducer {
+public:
+  explicit SiStripApvSimulationParametersESProducer(const edm::ParameterSet& conf);
+  ~SiStripApvSimulationParametersESProducer() override {}
+
+  std::unique_ptr<SiStripApvSimulationParameters> produce(const SiStripApvSimulationParametersRcd& record);
+
+private:
+  std::vector<edm::FileInPath> baselineFiles_TOB_;
+  std::vector<edm::FileInPath> baselineFiles_TIB_;
+  unsigned int baseline_nBins_;
+  float baseline_min_;
+  float baseline_max_;
+  std::vector<float> puBinEdges_;
+  std::vector<float> zBinEdges_;
+
+  SiStripApvSimulationParameters::LayerParameters makeLayerParameters(const std::string& apvBaselinesFileName) const;
+};
+
+SiStripApvSimulationParametersESProducer::SiStripApvSimulationParametersESProducer(const edm::ParameterSet& conf)
+    : baseline_nBins_(conf.getParameter<unsigned int>("apvBaselines_nBinsPerBaseline")),
+      baseline_min_(conf.getParameter<double>("apvBaselines_minBaseline")),
+      baseline_max_(conf.getParameter<double>("apvBaselines_maxBaseline")) {
+  setWhatProduced(this);
+  for (const auto x : conf.getParameter<std::vector<double>>("apvBaselines_puBinEdges")) {
```

Comment:
HI @EmyrClement @pieterdavid I've been trying to produce an actual payload  in sqlite format using the changes proposed here, but I am wondering how these two parameters should be configured:
  * `apvBaselines_puBinEdges`
  * `apvBaselines_zBinEdges`

in the original PR those two parameters are an empty vector: https://github.com/cms-sw/cmssw/pull/27823/files#diff-4fccbcd3f6c6635e96da2d43e86fc03aR128.
Am I missing something trivial?

Reply:
From looking into the file from which the ascii are derived, I guessed
```python
    apvBaselines_puBinEdges=cms.vdouble(0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18., 20., 22., 24., 26., 28., 30., 32., 34., 36., 38., 40., 42., 44., 46., 48., 50.),
    apvBaselines_zBinEdges=cms.vdouble(0., 25., 50., 75.),
```
I was trying the same (just pushed a few more commits), but got stuck just before with
```
Begin processing the 1st record. Run 1, Event 1, LumiSection 1 on stream 0 at 04-Sep-2019 19:20:15.281 CEST
----- Begin Fatal Exception 04-Sep-2019 19:20:15 CEST-----------------------
An exception of category 'NoRecord' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'p'
   [2] Calling method for module SiStripApvSimulationParametersBuilder/'prod'
Exception Message:
No "SiStripApvSimulationParametersRcd" record found in the EventSetup.n
 The Record is delivered by an ESSource or ESProducer but there is no valid IOV for the synchronizatio value.
 Please check 
   a) if the synchronization value is reasonable and report to the hypernews if it is not.
   b) else check that all ESSources have been properly configured. 
----- End Fatal Exception -------------------------------------------------
```
so I must have missed something in the ESProducer, I'll debug further tomorrow

Reply:
Here's the missing configuration (missing from original PR):
```
	process.SiStripSimBlock.apvBaselines_zBinEdges =  cms.vdouble( 0, 25, 50, 75 )
	process.SiStripSimBlock.apvBaselines_puBinEdges =  cms.vdouble( 0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50 )
```
There are len(apvBaselines_zBinEdges) (and len(apvBaselines_puBinEdges) ) bins in total, as there is an additional "overflow" bin i.e. for PU > 50 and z > 75.

Reply:
Thanks @EmyrClement, your configuration matches the one provided by Pieter at https://github.com/EmyrClement/cmssw/pull/3/files#r320884730.
If these parameters can be deduced by the input ASCII files, why one has to pass them from configuration? Are they somehow redundant?
At any rate, I took a slightly different route by writing a DB file writer that is basically a copy of the `ESProducer` here: https://github.com/cms-sw/cmssw/commit/5e39b8998775515ff8cc35c83074a40aa663be1b to overcome the issue with the `ESProducer` not working properly. This gets me past of the error reported in https://github.com/EmyrClement/cmssw/pull/3/files#r320884730 but  I stumbled in another issue:
```
----- Begin Fatal Exception 04-Sep-2019 21:11:03 CEST-----------------------
An exception of category 'RangeError' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'p'
   [2] Calling method for module SiStripApvSimulationParametersBuilder/'WriteInDB'
Exception Message:
Histogram3D bin 3645 out of range [0, 3644].
----- End Fatal Exception -------------------------------------------------

```

Reply:
The error above was due to a typo [here](https://github.com/EmyrClement/cmssw/pull/3/commits/2d99712343f99bdadcc319e81f7e0dd9c86650fb) - I also had a few off-by-one errors for the binning, but I think they went the other way.

I did the same, moving the ESProducer code into the "builder" EDAnalyzer, and now the DB part seems to be working: I could create an sqlite file, and it seems to be picked up by customising the digi step with
```python
    process.mix.digitizers.strip.includeAPVSimulation = True
    process.ApvSimParamsDB = cms.ESSource("PoolDBESSource",
            BlobStreamerName = cms.untracked.string('TBufferBlobStreamingService'),
            DBParameters = cms.PSet(authenticationPath = cms.untracked.string('/afs/cern.ch/cms/DB/conddb')),
            timetype = cms.string('runnumber'),
            connect = cms.string('sqlite_file:apvsimparameters.db'),
            toGet = cms.VPSet(cms.PSet(
                record = cms.string('SiStripApvSimulationParametersRcd'),
                tag = cms.string('SiStripApvSimulationParameters_2016preVFP_v1')
            )))
    process.es_prefer_ApvSimParams = cms.ESPrefer("PoolDBESSource", "ApvSimParamsDB")
```
now checking if the digitizer is actually using it, and if the payload is correct...

Reply:
> The error above was due to a typo here - I also had a few off-by-one errors for the binning, but I think they went the other way.

yeah, I noticed the same bug in the constructor.
Actually during the debugging I tried another one of the constructors, i.e.:

```
  SiStripApvSimulationParameters::LayerParameters layerParams(4,0.,75.,26,0.,50.,83,0.,728.);
```

instead of:
```
SiStripApvSimulationParameters::LayerParameters layerParams(zBinEdges_, puBinEdges_, baselineBinEdges);
```
and it worked (after fixing the off -by one) up to some point, but then the methods:
`upperLimitsX()`, `upperLimitsY()` and `upperLimitsZ()` were returning empty vectors.
I think the class implementation of `Histogram3D`is pretty messy...

Reply:
yes, it doesn't look like the most complete and clean code... using upperLimits everywhere was the lazy option, with the last commit it should also work for equidistant binnings.